### PR TITLE
fix(datastore): upgrade after destructive schema change

### PR DESCRIPTION
**Pull request by iartemiev**
#### Checklist

- [ ] PR description included
- [x] `yarn test` passes
- [ ] Tests are [changed or added](https://github.com/aws-amplify/amplify-js/blob/main/CONTRIBUTING.md#steps-towards-contributions)
- [ ] Relevant documentation is changed or added (and PR referenced)

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

File: packages/datastore/__tests__/indexeddb.upgrade.test.ts [hunk 1]
```diff
@@ -0,0 +1,121 @@
...
+		}
+
+		const exportedJSON = await readBlob(exportedBlob);
+		const exported = JSON.parse(exportedJSON);
+
+		for (const { schema } of exported.data.tables) {
+			expect(schema.split(',')).toContain('&[id]');
+		}
+
+		expect(exported).toMatchSnapshot(`v${DB_VERSION}-destructive-schema`);
```
Comment (svidgen) [inline packages/datastore/__tests__/indexeddb.upgrade.test.ts:119]: Checking indexedDB is OK. But, can we ensure DataStore is actually operational against the database post-upgrade?

File: packages/datastore/src/storage/adapter/IndexedDBAdapter.ts [hunk 2]
```diff
@@ -182,6 +182,16 @@ class IndexedDBAdapter implements Adapter {
...
 
+									const storeExists = !!Object.values(
+										this.schema.namespaces[namespaceName].models
+									)?.find(m => m.name === modelName);
+
+									if (!storeExists) {
+										// delete original
+										db.deleteObjectStore(tmpName);
+										continue;
+									}
```
Comment (manueliglesias) [inline packages/datastore/src/storage/adapter/IndexedDBAdapter.ts:193]:
```suggestion
									const storeExists = !!this.schema.namespaces[namespaceName].models[modelName];

									if (!storeExists) {
										// delete original
										db.deleteObjectStore(origStore.name);
										continue;
									}
```

File: packages/datastore/__tests__/indexeddb.upgrade.test.ts [hunk 3]
```diff
@@ -0,0 +1,121 @@
...
+
+	test(`Migration from v1 to v${DB_VERSION}`, async () => {
+		const v1Data = require('./v1schema.data.json');
+
+		const blob = new Blob([JSON.stringify(v1Data)], {
+			type: 'application/json',
+		});
+
+		// Import V1
+		(await Dexie.import(blob)).close();
```
Comment (svidgen) [inline packages/datastore/__tests__/indexeddb.upgrade.test.ts:85]: I understand that you TDD'd the fix, and maybe this is just my ignorance, but is there a sanity check we can add here to ensure Dexie loaded what we expect into indexedDB?

E.g., if someone starts adding tests and "tweaks" the dexie DI at the top, is there any possibility this test will test something *other* than what we expect it to?

File: packages/datastore/src/storage/adapter/IndexedDBAdapter.ts (hunk 2 above)
Comment (manueliglesias) [reply packages/datastore/src/storage/adapter/IndexedDBAdapter.ts:193]: - `this.schema.namespaces[namespaceName].models`'s keys are the model names, no need to extract the values and iterate over them.
- `db.deleteObjectStore(origStore.name);` is IMO more clear in its intention

Comment (manueliglesias) [reply packages/datastore/src/storage/adapter/IndexedDBAdapter.ts:193]: `const storeExists = modelName in this.schema.namespaces[namespaceName].models;` would work too

**Review by elorzafe — APPROVED** [review]
LGTM, 

maybe @manueliglesias suggestion are worth to use to make it more clear, what do you think @iartemiev

**Review by manueliglesias — DISMISSED** [review]
LGTM

**Review by svidgen — DISMISSED** [review]
Agree with @manueliglesias's comments. And, I think I'd like to see some extra protection against erosion of the test in the future. But, it looks like it'll do the job. If addressing the asks are very time-consuming, I'm favor shipping sooner rather than later on this one.

File: packages/datastore/__tests__/indexeddb.upgrade.test.ts (hunk 1 above)
Comment (iartemiev) [reply packages/datastore/__tests__/indexeddb.upgrade.test.ts:119]: Yea, can get by one of the items from the v1 store

File: packages/datastore/__tests__/indexeddb.upgrade.test.ts (hunk 3 above)
Comment (iartemiev) [reply packages/datastore/__tests__/indexeddb.upgrade.test.ts:85]: If we add tests to this, we'll want to do some cleanup in a before/after. I'll add a comment for future reference

File: packages/datastore/src/storage/adapter/IndexedDBAdapter.ts (hunk 2 above)
Comment (iartemiev) [reply packages/datastore/src/storage/adapter/IndexedDBAdapter.ts]: Yep, I ended up doing the `modelName in` approach - updated

**Review by elorzafe — APPROVED** [review]
Thanks @iartemiev

**Review by manueliglesias — APPROVED** [review]
LGTM 👍

**Pull request merged**